src/test/java/com/microsoft/azure/relay/AsyncLockTest.java

```diff
@@ -119,4 +121,91 @@ public void lockTimeoutTest() throws InterruptedException, ExecutionException, T
 		LockRelease lockRelease3 = lockFuture3.get(2000, TimeUnit.MILLISECONDS);
 		lockRelease3.release();
 	}
+	
+	@Test
+	public void lockScopeUncontended() throws Exception {
+		String originalString = "lockScopeUncontended";
+		AsyncLock lock = new AsyncLock();
+		assertFalse("Should not be locked to start", lock.isLocked());
+		
+		CompletableFuture<String> innerTask = new CompletableFuture<String>();
+		CompletableFuture<String> outerTask = lock.runInsideLockAsync(null, EXECUTOR, () -> {
+			assertTrue("Should be locked during Inner Function", lock.isLocked());
+			return innerTask;
+		});
+		
+		assertFalse(outerTask.isDone());
+		assertTrue("Should be locked now", lock.isLocked());
+		innerTask.complete(originalString);
+		String resultString = outerTask.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+		assertEquals(originalString, resultString);
+		assertFalse(lock.isLocked());
+	}
+	
+	@Test
+	public void lockScopeContended() throws Throwable {
+		String originalString = "lockScopeContended";
+		AsyncLock lock = new AsyncLock();
+		LockRelease lockRelease = lock.acquireAsync(EXECUTOR).join();
+		assertTrue("Should be locked now", lock.isLocked());
+		
+		CompletableFuture<Void> innerTaskStart = new CompletableFuture<Void>();
+		CompletableFuture<String> innerTaskResult = new CompletableFuture<String>();
+		CompletableFuture<String> outerTask = lock.runInsideLockAsync(null, EXECUTOR, () -> {
+			assertTrue("Should be locked during Inner Function", lock.isLocked());
+			innerTaskStart.complete(null);			
+			return innerTaskResult;
+		});
+		
+		assertFalse(outerTask.isDone());
+		assertTrue("Should be locked now", lock.isLocked());
+		innerTaskResult.complete(originalString);
```

Comment:
>innerTaskResult.complete(originalString); [](start = 2, length = 41)

So what's the purpose of completing it here? We can just return a completed future from inside directly. I'm not sure if instead of manually completing it every time, there will be more value in replacing it with a little function that sleeps or something to resemble more real life scenarios.

Reply:
I wanted the test to create outerTask but not have innerTaskResult completed yet. But since this test blocks on the AsyncLock already being locked I could simplify this. Is that what you're asking for?

---
In reply to: [259058136](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259058136) [](ancestors = 259058136)

Reply:
Simplifying this in the next iteration.

---
In reply to: [259453912](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259453912) [](ancestors = 259453912,259058136)